### ISSUE TITLE
Errors running test_NCBI_qblast.py

### ISSUE THREAD

**Issue by svalqui** [biopython/biopython#2738]
### Setup

I am reporting a problem with Biopython version, Python version, and operating
system as follows:

```python
Python 3.6.9 (default, Nov  7 2019, 10:44:02) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys; print(sys.version)
3.6.9 (default, Nov  7 2019, 10:44:02) 
[GCC 8.3.0]
>>> import platform; print(platform.python_implementation()); print(platform.platform())
CPython
Linux-4.15.0-88-generic-x86_64-with-Ubuntu-18.04-bionic
>>> import Bio; print(Bio.__version__)
1.76.dev0
```

(*Please copy and run the above in your Python, and copy-and-paste the output*)

### Expected behaviour

Test to pass

### Actual behaviour

```
test_blastp_nr_actin (__main__.TestQblast) ... ok
test_discomegablast (__main__.TestQblast) ... ok
test_error_conditions (__main__.TestQblast)
Test if exceptions were properly handled. ... ok
test_orchid_est (__main__.TestQblast) ... ok
test_parse_qblast_ref_page (__main__.TestQblast) ... ok
test_pcr_primers (__main__.TestQblast) ... ok
test_short_query (__main__.TestQblast)
Test SHORT_QUERY_ADJUST parameter. ... FAIL

======================================================================
FAIL: test_short_query (__main__.TestQblast)
Test SHORT_QUERY_ADJUST parameter.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_NCBI_qblast.py", line 224, in test_short_query
    self.assertEqual(len(my_hits.alignments), 5)
AssertionError: 0 != 5

----------------------------------------------------------------------
Ran 7 tests in 4183.082s

FAILED (failures=1)

```

### Steps to reproduce

```
~/PythonProjects/biopython/Tests$ python3 test_NCBI_qblast.py 
```

**Comment by peterjc** [comment]
Cross reference #2053 which added ``SHORT_QUERY_ADJUST`` support, perhaps @MarkusPiotrowski can take a look?

**Comment by MarkusPiotrowski** [comment]
There were some recent Python2/3 related changes and I fear that the online tests are often running 'under the radar' (already seen with `test_KEGG_online`).
I'll have a look.

**Comment by peterjc** [comment]
Indeed. Currently we only find out when someone tests locally and reports it, or in the run up to rolling a release (which is generally frustrating timing).

It would be fairly simple to make one of the TravisCI build targets include the online tests, but that seems a little excessive (in terms of load on other people's servers) and likely to cause false positives as they do fail intermittently at times.

Perhaps we could run them on TravisCI once a week in online mode, or something?

**Comment by MarkusPiotrowski** [comment]
Seems they changed the usual standard matrix for short protein queries from `PAM30` to `BLOSUM62`. They still use `PAM30` for their own online BLAST, but changed it for the remote service. We can't complain about this, because `SHORT_QUERY_ADJUST` is not officially supported.

**Comment by MarkusPiotrowski** [comment]
@svalqui I'm just curious, but for me also the test `test_parse_qblast_ref_page` fails. Are you sure that you run the test under Python 3? My failure seems to be related to the use of "StringIO".

**Comment by svalqui** [comment]
@MarkusPiotrowski , definitely Python3, I had some issues with timeouts because the test takes long, but for me it works as as initially logged; I have re-built, re-install to double check output below.

```
l:~/PythonProjects/biopython/Tests$ python3 test_NCBI_qblast.py 
test_blastp_nr_actin (__main__.TestQblast) ... ok
test_discomegablast (__main__.TestQblast) ... ok
test_error_conditions (__main__.TestQblast)
Test if exceptions were properly handled. ... ok
test_orchid_est (__main__.TestQblast) ... ok
test_parse_qblast_ref_page (__main__.TestQblast) ... ok
test_pcr_primers (__main__.TestQblast) ... ok
test_short_query (__main__.TestQblast)
Test SHORT_QUERY_ADJUST parameter. ... FAIL

======================================================================
FAIL: test_short_query (__main__.TestQblast)
Test SHORT_QUERY_ADJUST parameter.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_NCBI_qblast.py", line 301, in test_short_query
    self.assertEqual(len(my_hits.alignments), 5)
AssertionError: 0 != 5

----------------------------------------------------------------------
Ran 7 tests in 3763.516s

FAILED (failures=1)
~/PythonProjects/biopython/Tests$ 

``` 

If SHORT_QUERY_ADJUST is not officially supported, maybe it should be removed?.
Thank you for having a look at this.

**Comment by MarkusPiotrowski** [comment]
Oh, you ran the test within the test folder without using the test runner (`run_tests.py`). So the system will use your standard `Biopython`.
Can you try (within the test folder): `python run_tests.py test_NCBI_qblast`?

**Comment by peterjc** [comment]
Should be 'fixed' via #2742 but the SHORT_QUERY_ADJUST test in particular seems to take *hours*. That's not good for us waiting for it, but unless the nature of the computation changed (and the example is no longer trivial meaning it is also wasting NCBI compute time), what is causing the long delay?

**Comment by MarkusPiotrowski** [comment]
I thought the long delay is due to the NCBI penalizing us for too many requests in a short time. And the test for SHORT_QUERY_ADJUST is the last to run, so it took the most time. I wonder what the basis for a penalty may be, the IP address or the user-agent? In the last case this would be unfortunate, because all Biopython usage of `Bio.Blast.NCBIWWW` has the same user-agent: BiopythonClient.

**Comment by svalqui** [comment]
running ```python run_tests.py test_NCBI_qblast``` is taking me more than 2 hours, I'll try tomorrow.

**Comment by svalqui** [comment]
Ok, this is working now, Thank you, but it takes about 2 Hours

```
:~/PythonProjects/biopython/Tests$ python3 run_tests.py test_NCBI_qblast.py 
Python version: 3.6.9 (default, Nov  7 2019, 10:44:02) 
[GCC 8.3.0]
Operating system: posix linux
test_NCBI_qblast ... ok
----------------------------------------------------------------------
Ran 1 test in 7924.249 seconds
```
I have tried with different IP address, no difference; and I tried adding a sleep time of 60 seconds before every query and it didn't help either. It seems as the queries are send to a slower queue, not sure what is the criteria.

**Comment by peterjc** [comment]
We should try changing the sub-test order to confirm if ``SHORT_QUERY_ADJUST`` is slow, or simply it is an order effect. This is based on alphabetical sorting of the test method names - so can be manipulated fairly simply.

**Comment by svalqui** [comment]
Here is a break down...
```
~/PythonProjects/biopython/Tests$ python3 test_NCBI_qblast.py 
test_short_query (__main__.TestQblast)
Test SHORT_QUERY_ADJUST parameter. ... ok

----------------------------------------------------------------------
Ran 1 test in 823.603s

OK
```

```
:~/PythonProjects/biopython/Tests$ python3 test_NCBI_qblast.py 
test_pcr_primers (__main__.TestQblast) ... ok

----------------------------------------------------------------------
Ran 1 test in 63.374s

OK
```

```
~/PythonProjects/biopython/Tests$ python3 test_NCBI_qblast.py 
test_parse_qblast_ref_page (__main__.TestQblast) ... ok

----------------------------------------------------------------------
Ran 1 test in 0.000s

OK
```

```
:~/PythonProjects/biopython/Tests$ python3 test_NCBI_qblast.py 
test_orchid_est (__main__.TestQblast) ... ok

----------------------------------------------------------------------
Ran 1 test in 122.817s

OK
```

```
:~/PythonProjects/biopython/Tests$ python3 test_NCBI_qblast.py 
test_error_conditions (__main__.TestQblast)
Test if exceptions were properly handled. ... ok

----------------------------------------------------------------------
Ran 1 test in 0.000s

OK
```

```
:~/PythonProjects/biopython/Tests$ python3 test_NCBI_qblast.py 
test_discomegablast (__main__.TestQblast) ... ok

----------------------------------------------------------------------
Ran 1 test in 64.877s

OK
```

```
:~/PythonProjects/biopython/Tests$ python3 test_NCBI_qblast.py 
test_blastp_nr_actin (__main__.TestQblast) ... ok

----------------------------------------------------------------------
Ran 1 test in 2343.000s

OK
```
... but times can change significantly.

```
:~/PythonProjects/biopython/Tests$ python3 test_NCBI_qblast.py 
test_blastp_nr_actin (__main__.TestQblast) ... ok

----------------------------------------------------------------------
Ran 1 test in 63.456s

OK
```

**Comment by MarkusPiotrowski** [comment]
We can, of course, remove the negative result test, which demonstrates that there is no alignment without `short_query=True`. I placed it as an indicator for changes in the blast services (if this test fails, the search has become more relaxed).

**Comment by svalqui** [comment]
Just run the short query test, and removed ```short_query=True``` from it; the test fails; and takes a while.
```
:~/PythonProjects/biopython/Tests$ python3 test_NCBI_qblast.py 
test_short_query (__main__.TestQblast)
Test SHORT_QUERY_ADJUST parameter. ... FAIL

======================================================================
FAIL: test_short_query (__main__.TestQblast)
Test SHORT_QUERY_ADJUST parameter.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_NCBI_qblast.py", line 301, in test_short_query
    self.assertEqual(len(my_hits.alignments), 5)
AssertionError: 0 != 5

----------------------------------------------------------------------
Ran 1 test in 2543.707s

FAILED (failures=1)
```

Not sure where to go from here.

**Comment by MarkusPiotrowski** [comment]
1. Failing of this test without the 'short_query=True` parameter is expected. What I meant above is that there is another test within this test method (the first one), which does not have the parameter and returns 0 alignments (as expected). There are reasons for running this test, but I may be dropped. However:
2. I also ran these test several time yesterday. I ran each of the three tests, which are within the `test_short_query` method, independently. In a bad case *each* of this three tests took about 45 minutes, in the best case *each* test required only one minute. So one cannot really blame the tests themselves. @svalqui also reported very different timings for `test_blastp_nr_actin` (between 1 and 40 minutes). Also, it seems unpredictable at which time of the day the speed would be best.
3. So it's good that we do not run the online test too often. However, I cannot see what *we* can do about the slow response times. If we want to run this test regularly (e.g. once a week), we have to prepare for long run times.
